compiler/src/org.graalvm.compiler.core.test/src/org/graalvm/compiler/core/test/ReassociationTest.java

```diff
@@ -0,0 +1,474 @@
+    }
+
+    public static long refSubAddLongSnippet() {
+        return 4L + (rndL1 - rndL2);
+    }
+
+    @Test
+    public void testMulLong() {
+        testReassociateConstant("testMulLongSnippet", "refMulLongSnippet");
+    }
+
+    public static long testMulLongSnippet() {
+        return rndL1 * (-3L * rndL2) * -5L;
+    }
+
+    public static long refMulLongSnippet() {
+        return (rndL1 * rndL2) * 15L;
+    }
+
+    @Test
+    public void testAndLong() {
+        testReassociateConstant("testAndLongSnippet", "refAndLongSnippet");
+    }
+
+    public static long testAndLongSnippet() {
+        return rndL1 & (3L & rndL2) & 2L;
+    }
+
+    public static long refAndLongSnippet() {
+        return (rndL1 & rndL2) & 2L;
+    }
+
+    @Test
+    public void testOrLong() {
+        testReassociateConstant("testOrLongSnippet", "refOrLongSnippet");
+    }
+
+    public static long testOrLongSnippet() {
+        return rndL1 | (3L | rndL2) | 4L;
+    }
+
+    public static long refOrLongSnippet() {
+        return (rndL1 | rndL2) | 7L;
+    }
+
+    @Test
+    public void testXorLong() {
+        testReassociateConstant("testXorLongSnippet", "refXorLongSnippet");
+    }
+
+    public static long testXorLongSnippet() {
+        return rndL1 ^ (3L ^ rndL2) ^ 1L;
+    }
+
+    public static long refXorLongSnippet() {
+        return (rndL1 ^ rndL2) ^ 2L;
+    }
+
+    // Re-association overflow tests.
+    @Test
+    public void testOverflow1() {
+        testReassociateConstant("testOverflow1Snippet", "refOverflow1Snippet");
+    }
+
+    public static int testOverflow1Snippet() {
+        return rnd1 + (rnd2 + Integer.MAX_VALUE) + Integer.MIN_VALUE;
+    }
+
+    public static int refOverflow1Snippet() {
+        return (rnd1 + rnd2) - 1;
+    }
+
+    @Test
+    public void testOverflow2() {
+        testReassociateConstant("testOverflow2Snippet", "refOverflow2Snippet");
+    }
+
+    public static long testOverflow2Snippet() {
+        return rndL1 + (rndL2 + Long.MAX_VALUE) - Long.MIN_VALUE;
+    }
+
+    public static long refOverflow2Snippet() {
+        return (rndL1 + rndL2) + (-1L);
+    }
+
+    @Test
+    public void testOverflow3() {
+        testReassociateConstant("testOverflow3Snippet", "refOverflow3Snippet");
+    }
+
+    public static int testOverflow3Snippet() {
+        return rnd1 + (rnd2 - Integer.MIN_VALUE) - 5;
+    }
+
+    public static int refOverflow3Snippet() {
+        return (rnd1 + rnd2) - 0x80000005;
+    }
+
+    @Test
+    public void testOverflow4() {
+        testReassociateConstant("testOverflow4Snippet", "refOverflow4Snippet");
+    }
+
+    public static long testOverflow4Snippet() {
+        return rndL1 * (Long.MIN_VALUE * rndL2) * (-3L);
+    }
+
+    public static long refOverflow4Snippet() {
+        return (rndL1 * rndL2) * (Long.MIN_VALUE * (-3L));
+    }
+
+    private void testReassociateConstant(String testMethod, String refMethod) {
+        test(testMethod);
```

Comment:
that's something I overlooked. While this is a nice way of testing the pure functionality of the phase I think we also want to test the application of the phase embedded in the regular compilation pipeline.
For the test cases above can you please add assertions for the most important node counts, e.g. for the mul test ensure after a regular compilation the final low tier graph before backend contains one mul node with constant 15 and so on.
We want to especially test the mul stuff we are discussing about and for that we need to test the final graph result after all phases.

Reply:
I think the final graph y added should be sufficient, thx. Verifying the node counts by hand would just guarantee that if we start doing something categorically wrong both inside the test and the ref snippet we would still catch it but thats not necessary for now.